app/models/manageiq/providers/native_crud.rb

```diff
@@ -0,0 +1,129 @@
+class ManageIQ::Providers::NativeCrud < Job
+  def self.create_job(options)
+    super(name, options)
+  end
+
+  #
+  # State-transition diagram:
+  #
+  #    *                          /------\
+  #    | :initialize              |      |  :poll_native_task
+  #    v               :start     v      |
+  # waiting_to_start  -------> running ----------> refreshing <-\
+  #                               |     :refresh        |       | :poll_refresh
+  #                               v                     |-------/
+  #                             error <-----------------|
+  #                               |                     |
+  #                               v                     |
+  #       finished <---------- notifying --------------/
+  #                   :finish              :notify
+  #
+  def load_transitions
+    self.state ||= 'initialize'
+
+    {
+      :initializing     => {'initialize'       => 'waiting_to_start'},
+      :start            => {'waiting_to_start' => 'running'},
+      :poll_native_task => {'running'          => 'running'},
+      :refresh          => {'running'          => 'refreshing'},
+      :poll_refresh     => {'refreshing'       => 'refreshing'},
+      :notify           => {'*'                => 'notifying'},
+      :finish           => {'*'                => 'finished'},
+      :abort_job        => {'*'                => 'aborting'},
+      :cancel           => {'*'                => 'canceling'},
+      :error            => {'*'                => '*'}
+    }
+  end
+
+  def run_native_op
+    raise NotImplementedError, _("run_native_op must be implemented by a subclass")
+  end
+
+  def poll_native_task
+    raise NotImplementedError, _("poll_native_task must be implemented by a subclass")
+  end
+
+  def refresh
+    target = target_entity
+
+    task_ids = EmsRefresh.queue_refresh_task(target)
+    if task_ids.blank?
+      queue_signal(:error, "Failed to queue refresh", "error")
+    else
+      context[:refresh_task_ids] = task_ids
+      update_attributes!(:context => context)
+
+      queue_signal(:poll_refresh)
+    end
+  end
+
+  def poll_refresh
+    refresh_finished = true
+
+    context[:refresh_task_ids].each do |task_id|
+      task = MiqTask.find(task_id)
+      if task.status != MiqTask::STATUS_OK
+        return queue_signal(:error, "Refresh failed", "error")
+      end
+
+      if task.state != MiqTask::STATE_FINISHED
+        refresh_finished = false
+        break
+      end
+    end
+
+    if refresh_finished
```

Comment:
> Can refresh_finished? become a private method that's called by poll_refresh?

Yeah I like that a lot, I'll just need to handle the refresh task error differently since right now it is returning out of poll_refresh and signaling `:error`